snippets/base/forms.py

```diff
@@ -534,3 +534,51 @@ def clean(self):
             if main_body_count > 1:
                 raise forms.ValidationError(
                     'There can be only one Main Text variable type per template')
+
+
+class ASRSnippetAdminForm(forms.ModelForm):
+    template = forms.ModelChoiceField(
+        queryset=SnippetTemplate.objects.exclude(hidden=True).filter(startpage__gte=6),
+        widget=TemplateSelect)
+
+    class Meta:
+        model = ASRSnippet
+        exclude = ['creator', 'created', 'modified']
+        widgets = {
+            'data': TemplateDataWidget('template'),
+        }
+
+
+class TargetAdminForm(forms.ModelForm):
+    filtr_is_default_browser = fields.JEXLChoiceField(
+        'isDefaultBrowser',
+        choices=((None, "I don't care"),
+                 ('true', 'Yes',),
+                 ('false', 'No')),
+        label_suffix='?',
+        label='Has Firefox set to be the default browser',
+        required=False)
+
+    class Meta:
+        model = Target
+        exclude = ['creator', 'created', 'modified']
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.instance and self.instance.jexl:
+            for name, field in self.fields.items():
+                if name.startswith('filtr_'):
+                    field.initial = self.instance.jexl.get(name[6:])
+
+    def save(self, *args, **kwargs):
+        jexl = {}
+        jexl_expr_array = []
+
+        for name, field in self.fields.items():
+            if name.startswith('filtr_'):
+                jexl[name[6:]] = self.cleaned_data[name]
+                jexl_expr_array.append(field.to_jexl(self.cleaned_data[name]))
+        self.instance.jexl = jexl
+        self.instance.jexl_expr = ' && '.join([x for x in jexl_expr_array if x])
```

Comment:
I like the dynamic field approach just fine. I do think that keeping the serialization of the jexl dynamic columns to `jexl_expr` in the `save()` method on the model makes the most sense as it could be confusing if this model were updated and saved someday outside of this form. I also think adding a `spec` to the `jexl` dynamic field definition is a good thing to do with these things since it's a bit more self-documenting and will ensure data integrity.

Reply:
Thanks for the input and good point on the spec, added.